### PR TITLE
Show the warning message in the case with deprecated plugins

### PR DESCRIPTION
**Pull request by olexii4**
Signed-off-by: Oleksii Orel <oorel@redhat.com>

### What does this PR do?
Add implementation of the warning message in the case with deprecated plugins and the possibility to auto-migrate plugins by pressing the button.

### What issues does this PR fix or reference?
Needed for https://github.com/eclipse/che/issues/13021

#### Release Notes

 - show a warning message in the case with deprecated editor or plugins.
 - add a possibility to migrate by pressing the button.

#### Docs PR

Screenshot from 2019-10-06 20-55-32
Screenshot from 2019-10-06 20-56-19

https://youtu.be/7U00YxALrFE

### CODE REVIEW

**Comment by olexii4** [comment]
@slemeur WDYT?

**Comment by olexii4** [comment]
ci-test

**Review by mmorhun — APPROVED** [review]

**Review by benoitf — APPROVED** [review]

**Review by AndrienkoAleksandr — APPROVED** [review]

**Review by vitaliy-guliy — APPROVED** [review]

**Comment by slemeur** [comment]
Is it migrating or upgrading ?

**Comment by olexii4** [comment]
I think it could be an upgrading in the case when we just upgrade the version of the same plugin. But it can be a migration too.

**Review by amisevsk — COMMENTED** [review]
Just to double-check, does the warning only appear on the Plugins tab, or whenever workspace details are opened?

**Comment by olexii4** [comment]
@amisevsk The warning only appears on the Plugins tab.

**Comment by amisevsk** [comment]
> The warning only appears on the Plugins tab.

I think I understand why that is the case, but also worry that a user might never see it (I don't know how often a user would visit the plugins/editors tab for an already set up workspace).

+1 on the changes, regardless.

**Comment by olexii4** [comment]
I have updated
Screenshot from 2019-10-03 20-23-26
Screenshot from 2019-10-03 20-45-44
Screenshot from 2019-10-03 20-44-46

https://youtu.be/qkjxKXLGZ70

**Comment by olexii4** [comment]
@slemeur, @amisevsk   WDYT?

**Comment by olexii4** [comment]
crw-ci-test

**Comment by olexii4** [comment]
crw-ci-test

**Comment by olexii4** [comment]
crw-ci-test

**Comment by slemeur** [comment]
Based on the video shared here: https://www.youtube.com/watch?v=qkjxKXLGZ70&feature=youtu.be

> The workspace includes deprecated plugins. Migrate deprecated plugins by pressing the button Migrate plugins.
The workspace uses deprecated plugins. Upgrade the plugins to their latest version by pressing the button *Upgrade plugins*. 

> This plugin is deprecated. Migrate to the next .... version by pressing the icon
This is unclear. Which icon user has to press? It's a warning icon, a user will not expect to be able to click on it to fix the problem. I would remove the action from the button, and just explain what to do.
Message: "This plugin is deprecated. Select a newer version in the dropdown list or press "Upgrade plugins" to upgrade all plugins to their latest version".

* For editors*
- The error message mention "Migrate plugins", but the button is not there. 
- Should the button be "Upgrade deprecated components"? 

> The workspace includes deprecated plugins and editors. Migrate deprecated plugins by pressing the button Migrate plugins.
The workspace uses deprecated plugins and editor. Upgrade to the latest versions available by pressing the button *Upgrade plugins*. 

> This workspace includes the deprecated editor
This workspaces uses a deprecated editor.

I find the experience a bit challenging and confusing for a end-user. Should the action be closer to the error message? Could you have "*Click Here* to upgrade all deprecated components" at the end of the error message, which will trigger the action ?


cc @evidolob .

**Comment by sunix** [comment]
We discussed during standup that https://github.com/eclipse/che/issues/13021 was just about showing a warning message about the deprecation of the plugin version. Not the feature to auto migrate that could introduce new bugs and has not been prioritised.

@olexii4 have you updated your PR based on our discussion ?

**Comment by olexii4** [comment]
ci-test

**Comment by olexii4** [comment]
I have updated
Screenshot from 2019-10-06 20-55-32
Screenshot from 2019-10-06 20-56-19

https://youtu.be/7U00YxALrFE

**Comment by olexii4** [comment]
@sunix I have  updated my PR based on our discussion

**Comment by olexii4** [comment]
crw-ci-test

**Comment by olexii4** [comment]
ci-test

**Comment by olexii4** [comment]
ci-build

**Comment by olexii4** [comment]
ci-test

**Comment by olexii4** [comment]
@slemeur Is it ok for you?

File: dashboard/yarn.lock [hunk 1]
```diff
@@ -700,13 +686,13 @@ aws4@^1.2.1, aws4@^1.8.0:
   resolved "https://registry.yarnpkg.com/aws4/-/aws4-1.8.0.tgz#f0e003d9ca9e7f59c7a508945d7b2ef9a04a542f"
   integrity sha512-ReZxvNHIOv88FlT7rxcXIIC0fPt4KZqZbOlivyWtXLt8ESx84zd3kMC6iK5jVeS2qt+g7ftS7ye4fi06X5rtRQ==
 
-axios@0.17.1:
-  version "0.17.1"
-  resolved "https://registry.yarnpkg.com/axios/-/axios-0.17.1.tgz#2d8e3e5d0bdbd7327f91bc814f5c57660f81824d"
-  integrity sha1-LY4+XQvb1zJ/kbyBT1xXZg+Bgk0=
+axios@0.19.0:
```
Comment (akurinnoy) [inline dashboard/yarn.lock]: Do you know why `yarn.lock` was changed?
btw, we don't have CQ for `axios@0.19.0`

Comment (olexii4) [reply dashboard/yarn.lock]: Autogenerated. Fixed.

File: dashboard/src/components/api/workspace/che-workspace.factory.ts [hunk 2]
```diff
@@ -181,6 +181,10 @@ export class CheWorkspace {
     this.checkWorkspaceLoader(userDashboardConfig.developmentMode, proxySettings);
   }
 
+  get hasWorkspaceAPI() {
```
[the remark this answers is not in the document]
Comment (olexii4) [reply dashboard/src/components/api/workspace/che-workspace.factory.ts]: removed

File: dashboard/src/app/workspaces/workspace-details/workspace-plugins/workspace-plugins.controller.ts [hunk 3]
```diff
@@ -169,23 +173,80 @@ export class WorkspacePluginsController {
...
+    if (deprecatedInfo && deprecatedInfo.migrateTo) {
+      const {publisher, name} = this.splitPluginId(pluginId);
+      const migrateToInfo = this.splitPluginId(deprecatedInfo.migrateTo);
+      if (migrateToInfo.publisher === publisher && migrateToInfo.name === name) {
+        warningMessage += ' Select a newer version in the dropdown list.';
       } else {
-        currentlySelectedPlugins.push(pluginIDWithVersion);
+        const targetPlugin = this.plugins.get(`${migrateToInfo.publisher}/${migrateToInfo.name}`);
+        if (targetPlugin) {
+          warningMessage += ` Select a newer version of <b>${targetPlugin.displayName} (${targetPlugin.publisher} publisher)</b>.`;
```
[the remark this answers is not in the document]
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/workspace-plugins/workspace-plugins.controller.ts]: Ok. Done

File: dashboard/src/app/workspaces/workspace-details/workspace-plugins/workspace-plugins.controller.ts [hunk 4]
```diff
@@ -169,23 +173,80 @@ export class WorkspacePluginsController {
...
+   *
+   * @param {string} pluginId
+   * @returns {any} warning message
+   */
+  getWarningMessage(pluginId: string): any {
+    let warningMessage = 'This plugin is deprecated.';
+    const deprecatedInfo = this.deprecatedPluginsInfo.get(pluginId);
+    if (deprecatedInfo && deprecatedInfo.migrateTo) {
+      const {publisher, name} = this.splitPluginId(pluginId);
+      const migrateToInfo = this.splitPluginId(deprecatedInfo.migrateTo);
```
[the remark this answers is not in the document]
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/workspace-plugins/workspace-plugins.controller.ts]: done

File: dashboard/src/app/workspaces/workspace-details/workspace-details.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  describe(`Get selected deprecated >`, () => {
+
+    it(`Plugins >`, () => {
+      const workspace = getWorkspaceDetails();
+
+      expect(workspaceDetailsService.getSelectedDeprecatedPlugins(workspace)).toEqual([]);
+
+      const {id} = getPlugins().find(item => item.deprecate && item.type !== PluginRegistry.EDITOR_TYPE);
+      const type = 'chePlugin';
+      workspace.devfile.components.push({id,type});
```
[the remark this answers is not in the document]
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/workspace-details.service.spec.ts]: done

File: dashboard/src/app/workspaces/workspace-details/workspace-details.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+   * Check assertion after the test
+   */
+  afterEach(() => {
+    $httpBackend.verifyNoOutstandingExpectation();
+    $httpBackend.verifyNoOutstandingRequest();
+  });
+
+  describe(`Get selected deprecated >`, () => {
+
+    it(`Plugins >`, () => {
```
[the remark this answers is not in the document]
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/workspace-details.service.spec.ts]: done

[the remark this answers is not in the document]
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/workspace-details.service.spec.ts]: done

**Review by akurinnoy — APPROVED** [review]
Thanks! Nice work, Oleksii!

**Review by evidolob — APPROVED** [review]

**Pull request merged**